File: src/Layout.php
<?php

declare(strict_types=1);

namespace MakeImageSquare;

final class Layout
{
    public const MODE_FIT = 'fit';
    public const MODE_FILL = 'fill';

    /**
     * @return array{
     *   canvas: float,
     *   contentWidth: float,
     *   contentHeight: float,
     *   offsetX: float,
     *   offsetY: float,
     *   padLeft: float,
     *   padRight: float,
     *   padTop: float,
     *   padBottom: float
     * }
     */
    public static function makeImageSquare(
        float $width,
        float $height,
        string $mode = self::MODE_FIT,
        ?float $canvasSize = null
    ): array {
        if (!is_finite($width) || !is_finite($height) || $width <= 0.0 || $height <= 0.0) {
            throw new \InvalidArgumentException('width and height must be positive numbers');
        }

        $canvas = $canvasSize !== null ? $canvasSize : max($width, $height);
        if (!is_finite($canvas) || $canvas <= 0.0) {
            throw new \InvalidArgumentException('canvas size must be a positive number');
        }

        if ($mode === self::MODE_FILL) {
            $scale = $canvas / min($width, $height);
            $contentWidth = $width * $scale;
            $contentHeight = $height * $scale;
            $offsetX = ($canvas - $contentWidth) / 2.0;
            $offsetY = ($canvas - $contentHeight) / 2.0;

            return [
                'canvas' => $canvas,
                'contentWidth' => $contentWidth,
                'contentHeight' => $contentHeight,
                'offsetX' => $offsetX,
                'offsetY' => $offsetY,
                'padLeft' => max(0.0, $offsetX),
                'padRight' => max(0.0, $canvas - $contentWidth - $offsetX),
                'padTop' => max(0.0, $offsetY),
                'padBottom' => max(0.0, $canvas - $contentHeight - $offsetY),
            ];
        }

        if ($mode !== self::MODE_FIT) {
            throw new \InvalidArgumentException("mode must be 'fit' or 'fill'");
        }

        $scale = $canvas / max($width, $height);
        $contentWidth = $width * $scale;
        $contentHeight = $height * $scale;
        $padX = ($canvas - $contentWidth) / 2.0;
        $padY = ($canvas - $contentHeight) / 2.0;

        return [
            'canvas' => $canvas,
            'contentWidth' => $contentWidth,
            'contentHeight' => $contentHeight,
            'offsetX' => $padX,
            'offsetY' => $padY,
            'padLeft' => $padX,
            'padRight' => $padX,
            'padTop' => $padY,
            'padBottom' => $padY,
        ];
    }

    /**
     * @return array{canvas: float, left: float, right: float, top: float, bottom: float}
     */
    public static function squarePadding(float $width, float $height, ?float $canvasSize = null): array
    {
        $layout = self::makeImageSquare($width, $height, self::MODE_FIT, $canvasSize);

        return [
            'canvas' => $layout['canvas'],
            'left' => $layout['padLeft'],
            'right' => $layout['padRight'],
            'top' => $layout['padTop'],
            'bottom' => $layout['padBottom'],
        ];
    }
}

File: src/main/java/io/github/ilovejackylee/makeimagesquare/Layout.java
package io.github.ilovejackylee.makeimagesquare;

/**
 * Layout math for fitting an image into a 1:1 canvas.
 * Does not decode or rewrite image files.
 * <p>
 * Prefer a UI instead of calling this from Java? There is a
 * <a href="https://squareimage.run">pad-to-square editor that runs in the browser</a>.
 */
public final class Layout {
    private Layout() {
    }

    public static SquareLayout makeImageSquare(double width, double height) {
        return makeImageSquare(width, height, SquareMode.FIT, null);
    }

    public static SquareLayout makeImageSquare(
            double width, double height, SquareMode mode, Double canvasSize) {
        requirePositive(width, "width");
        requirePositive(height, "height");
        double canvas = canvasSize != null ? canvasSize : Math.max(width, height);
        requirePositive(canvas, "canvasSize");

        if (mode == SquareMode.FILL) {
            double scale = canvas / Math.min(width, height);
            double contentWidth = width * scale;
            double contentHeight = height * scale;
            double offsetX = (canvas - contentWidth) / 2.0;
            double offsetY = (canvas - contentHeight) / 2.0;
            return new SquareLayout(
                    canvas,
                    contentWidth,
                    contentHeight,
                    offsetX,
                    offsetY,
                    Math.max(0, offsetX),
                    Math.max(0, canvas - contentWidth - offsetX),
                    Math.max(0, offsetY),
                    Math.max(0, canvas - contentHeight - offsetY));
        }

        double scale = canvas / Math.max(width, height);
        double contentWidth = width * scale;
        double contentHeight = height * scale;
        double padX = (canvas - contentWidth) / 2.0;
        double padY = (canvas - contentHeight) / 2.0;
        return new SquareLayout(
                canvas,
                contentWidth,
                contentHeight,
                padX,
                padY,
                padX,
                padX,
                padY,
                padY);
    }

    public static SquarePadding padding(double width, double height) {
        return padding(width, height, null);
    }

    public static SquarePadding padding(double width, double height, Double canvasSize) {
        SquareLayout layout = makeImageSquare(width, height, SquareMode.FIT, canvasSize);
        return new SquarePadding(layout.canvas, layout.padLeft, layout.padRight, layout.padTop, layout.padBottom);
    }

    private static void requirePositive(double value, String name) {
        if (!Double.isFinite(value) || value <= 0) {
            throw new IllegalArgumentException(name + " must be a positive number");
        }
    }
}
